### ISSUE TITLE
Naming conventions and organization

### ISSUE THREAD

**Issue by jamescheney** [links-lang/links#547]
Links source files include things like:

```
desugarCP.ml
sugartypes.ml
typeSugar.ml
irCheck.ml
evalir.ml
webserver_types.ml
parseXml.mly
jsonlex.mll
jsonparse.mly
```

Some observations that make me sad are:

- Names like `sugartypes` and `typeSugar` provide almost no clue as to what these modules do, why there are two of them and how they differ.  `sugartypes` contains type definitions and some utility functions for the "Sugar AST", i.e. the frontend AST, while `typeSugar` typechecks this AST.  
- Meanwhile, `irCheck` checks the IR.  So, some module names are of the form "noun verb" (`irCheck`,`jsonlex`) while others are of the form "verb noun" (`typeSugar`, `parseXml`) and still others are "noun noun" (`sugartypes`).  Of course there won't be a single best way to do this but perhaps we can cut down on the number of ways we use.
- `sugar*` and `desugar*` are not intuitive module names to me.  Usually `sugar*` means "something to do with the sugar AST" (but could also in this context be about performing sugaring) while `desugar*` means "module that does a desugaring / elaboration stage on the sugar AST".  I think it would be simpler to call the frontend AST something like `absyn` (a noun) and replace all the uses of (de)sugar with something more descriptive.  
- We adhere to two or three conventions regarding multi-word module/filenames or abbreviations (camelCase, snake_case, runoncase; some acronyms are fully capitalized (CP), others not (Xml, Json).  
- Now that we are starting to use subdirectories to collect certain subsystems (lens, query) maybe there is an opportunity to further decompose (sugar/frontent/absyn vs. backend/IR vs. webby stuff) - and perhaps move more of the database stuff into `query` too.

My proposal:
- Adopt CamelCase or snake_case everywhere.  The latter seems a bit more standard in OCaml but would involve a lot more changes since we mostly use CamelCase
- Standardize on capitalizing acronyms entirely.  Acronyms should only appear at the end of a module name (i.e. it should be `parseJSON` not `JSONParse`
- Prefer "verb noun" for modules that do something, and "noun noun" for modules that provide something, e.g. `typecheckIR`, `typecheckAbsyn` vs `absynTypes`, `absynTraversals`
- Wherever it makes sense, plunge related modules down a directory level

I'm sure each of these suggestions will be controversial.  Discuss!

**Comment by jstolarek** [comment]
> I'm sure each of these suggestions will be controversial. 

I actually like your proposals. Except maybe for `absyn`. `LinksSyn` maybe?

**Comment by rudihorn** [comment]
I mostly am also happy to have these things better defined. It would probably make sense to discuss this at the links meeting.

**Comment by frank-emrich** [comment]
+1 for this
Trying to remember which file/module names use camel case and which don't is one of those daily annoyances.

In general, this is one of those instances where the absence of an implicit convention in the existing code prevents people from just sticking to it. This is why we should agree on naming conventions, etc, change all violations in the codebase, and enforce sticking to the convention from then on.

**Comment by jstolarek** [comment]
One more thought. I have a preference for capitalizing file names, so `TypecheckIR` and not `typecheckIR`. This way the name of the file is identical with the name of a module the file creates.

**Comment by frank-emrich** [comment]
> I have a preference for capitalizing file names

It looks like all major projects I've seen so far and all OCaml documentation uses lower-case file names. However, I wasn't able to find any style guide (like https://ocaml.org/learn/tutorials/guidelines.html) that explicitly states this as a convention. Of course, this somewhat prevents us from starting file names with acronyms if we want to capitalize them (which I like).

**Comment by jamescheney** [comment]
Yes, that's why I suggested the convention of capitalizing acronyms but requiring them to be at the end.  Alternatively, allowing acronyms at the beginning of a camelcase file name to be all-lowercase would also work, but seems perverse (xmlParse rather than `XMLParse` `xMLParse`).  

I lean a bit away from capitalizing file names, perhaps just because I have internalized the "foo.ml" ->  "open Foo" convention.

Since it sounds like there's some consensus forming around this, but we haven't heard from everyone, I suggest we proceed as follows:

- I'll put up a wiki page on naming conventions, with a giant code block listing all of the current (core) filenames
- I'll suggest a rationalization of the names along the lines sketched above which we can discuss on Thursday.

**Comment by jamescheney** [comment]
OK, so the page is here.  Not surprisingly, the simple rules I came up with don't handle all the current cases well, so I'm willing to revisit adopting snake_case or CamelCase with only the first letter of acroynms capitalized, e.g. `parseXml` rather than `parseXML` (maybe treating desugarCP as a special case).

https://github.com/links-lang/links/wiki/Naming-conventions

**Comment by jstolarek** [comment]
@jamescheney, what would be the best way to comment on your proposed name changes? Should I write my comments here or update the wiki somehow?

**Comment by jamescheney** [comment]
I was thinking adding comments to the wiki (below what I wrote) would be easiest.  An alternative approach could be to create a pr (not intended to be merged) with the proposed renamings in a text file, then people could comment line by line, but not sure that's needed yet. I can do that if people think it'd be better.

**Comment by jamescheney** [comment]
In today's meeting we discussed the desirability of also:
-  [ ] adding guidelines / reminders about how to document things (e.g. banner comments for modules/interfaces) to the "naming conventions" page (generalized to "naming and documentation conventions)
-  [ ] adding a file with the proposed renamings as a PR, which we can discuss/agree.  Then once we are happy, actually add this to the repository, as documentation of how things have been renamed, since past issues might still have the old names.

I will create a PR and add tentative documentation conventions to the wiki page.

**Comment by jstolarek** [comment]
Regarding the wiki, if some behaviour is documented on the wiki it is a good idea to put a reminder in the source code to update the wiki shall the behaviour change. It is an even better idea to actually read these comments and update the wiki when needed! Sadly, this doesn't always happen, as exemplified by the `run-tests` script:

```
# If you modify the behaviour of this file please remember to update the
# relevant wiki page:
#
# https://github.com/links-lang/links/wiki/Running-the-test-suite
```
The wiki was not updated to reflect the addition of unit tests :(

**Comment by jamescheney** [comment]
Perhaps we can have a script that checks for this:

```
IF file XYZ contains a machine-readable link to a wiki page ABC
AND  the last modified date of ABC < the last modified data of XYZ
THEN complain 
```
True, this can be got around by just doing `touch ABC` but at least it would be harder to ignore as a reminder...
